### PR TITLE
Make ServoParser::pending_input hold onto a BufferQueue

### PR DESCRIPTION
**Pull request by nox**
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/14250)

### CODE REVIEW

**Comment by SimonSapin** [comment]
Reviewed 6 of 6 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[components/script/dom/servoparser/xml.rs, line 54 at r1](https://reviewable.io:443/reviews/servo/servo/14250#-KWhot2rU-Dy9RwgUA_0:-KWhot2sZAdX83HBaTo5:b52lv7i) ([raw file](https://github.com/servo/servo/blob/cc8de1d7814efb62bb26283e56378542d3c6bb0c/components/script/dom/servoparser/xml.rs#L54)):*
> ```Rust
> 
>     pub fn feed(&mut self, input: &mut BufferQueue) {
>         if let Some(chunk) = input.pop_front() {
> ```

Should this be a loop? r=me otherwise

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/14250)*

File: components/script/dom/servoparser/xml.rs [hunk 1]
```diff
@@ -165,13 +183,7 @@ impl<'a> TreeSink for Sink {
...
     fn complete_script(&mut self, node: Self::Handle) -> NextParserState {
-        let script = node.downcast::<HTMLScriptElement>();
-        if let Some(script) = script {
-            return match script.prepare() {
-                true => NextParserState::Continue,
-                false => NextParserState::Suspend,
-            };
-        }
-        NextParserState::Continue
+        self.script.set(Some(node.downcast().unwrap()));
```
Comment (nox) [inline components/script/dom/servoparser/xml.rs]: This is the only way I found to make xml5ever just stop tokenizing on `</script>` to handle the script itself out of it.

**Comment by nox** [comment]
```
  ▶ CRASH [expected TIMEOUT] /html/semantics/scripting-1/the-script-element/script-text-xhtml.xhtml
  │ 
  │ VMware, Inc.
  │ Gallium 0.4 on softpipe
  │ 3.3 (Core Profile) Mesa 12.0.1
  │ called `Option::unwrap()` on a `None` value (thread ScriptThread PipelineId { namespace_id: PipelineNamespaceId(0), index: PipelineIndex(1) }, at ../src/libcore/option.rs:323)
  │ stack backtrace:
  │    0:        0x10e6df1de - backtrace::backtrace::trace::h16372ee7bf1517e5
  │    1:        0x10e6df4cc - backtrace::capture::Backtrace::new::hcc43c50c4b11c693
  │    2:        0x10d571e44 - servo::main::_{{closure}}::he01100fc53982b46
  │    3:        0x10ef8d923 - std::panicking::rust_panic_with_hook::h00b81bb3dcbd51f2
  │    4:        0x10ef8d7f4 - std::panicking::begin_panic::ha6a0d553db9869ff
  │    5:        0x10ef8d712 - std::panicking::begin_panic_fmt::h24d113aee3ee4081
  │    6:        0x10ef8d677 - rust_begin_unwind
  │    7:        0x10efba440 - core::panicking::panic_fmt::he441b2ea2036b98a
  │    8:        0x10efba344 - core::panicking::panic::haf296e94ad32f436
  │    9:        0x10dc5125b - _<script..dom..servoparser..xml..Sink as xml5ever..tree_builder..interface..TreeSink>::complete_script::h363a22d4574fe967
  │   10:        0x10d7a8305 - _<xml5ever..tree_builder..XmlTreeBuilder<Handle, Sink> as xml5ever..tokenizer..interface..TokenSink>::process_token::h5352793fa63d0833
  │   11:        0x10da2414d - _<xml5ever..tokenizer..XmlTokenizer<Sink>>::process_token::h298bec2821ee1492
  │   12:        0x10da246ef - _<xml5ever..tokenizer..XmlTokenizer<Sink>>::emit_current_tag::hca225c6642ecfea1
  │   13:        0x10da2739d - _<xml5ever..tokenizer..XmlTokenizer<Sink>>::step::h8baab31b871abdc7
  │   14:        0x10da259e7 - _<xml5ever..tokenizer..XmlTokenizer<Sink>>::run::h87e3b4d6a1824607
  │   15:        0x10dc53f2e - script::dom::servoparser::ServoParser::parse_sync::h6edc19833d00e6a4
  │   16:        0x10db1c6e5 - script::dom::document::Document::finish_load::hf165040e1edb94ab
  │   17:        0x10dbd449b - _<script..dom..htmlscriptelement..ScriptContext as net_traits..FetchResponseListener>::process_response_eof::h86f91cedb9e1b62a
  │   18:        0x10dcd4b5a - _<script..network_listener..ListenerRunnable<A, Listener> as script..script_thread..Runnable>::handler::h63993bd2bbce91fa
  │   19:        0x10dcdf472 - _<script..script_thread..CancellableRunnable<T> as script..script_thread..Runnable>::handler::h74f55f00d39311c3
  │   20:        0x10dcf8451 - script::script_thread::ScriptThread::handle_msg_from_script::h4a8186d0d2190624
  │   21:        0x10dea612c - script::script_thread::ScriptThread::handle_msgs::_{{closure}}::heac91dc058b267c7
  │   22:        0x10dceef6c - script::script_thread::ScriptThread::handle_msgs::hef5c3f0cf849f64e
  │   23:        0x10d846e37 - std::panicking::try::do_call::h273505ea6976b345
  │   24:        0x10ef8e94a - __rust_maybe_catch_panic
  │   25:        0x10da0f936 - _<F as alloc..boxed..FnBox<A>>::call_box::hfed4a82f33ee708a
  │   26:        0x10ef8c7d4 - std::sys::thread::Thread::new::thread_start::h9fa66f6812e81e10
  │   27:     0x7fff89039059 - _pthread_body
  │   28:     0x7fff89038fd6 - _pthread_start
  │ ERROR:servo: called `Option::unwrap()` on a `None` value
  └ Pipeline failed in hard-fail mode.  Crashing!
```

So apparently xml5ever sometimes calls `TreeSink::complete_script` with nodes that aren't scripts, great...

**Comment by SimonSapin** [comment]
@bors-servo r+

---

Reviewed 1 of 7 files at r2, 6 of 6 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/14250#-:-KWrhIh8DnqI7UjO8wKQ:bkg9nno)*

**Pull request merged**